drivers/rc/lirc_dev.c

```diff
@@ -0,0 +1,1174 @@
+ *   ev     - The data of receiving from IR device
+ ****************************************************************************/
+
+void lirc_raw_event(FAR struct lirc_lowerhalf_s *lower,
+                    struct lirc_raw_event_s ev)
+{
+  FAR struct lirc_upperhalf_s *upper = lower->priv;
+  FAR struct list_node *node;
+  FAR struct list_node *tmp;
+  FAR struct lirc_fh_s *fh;
+  irqstate_t flags;
+  int semcount;
+  int sample;
+  int gap;
+
+  /* Packet start */
+
+  if (ev.reset)
+    {
+      /* Userspace expects a long space event before the start of
+       * the signal to use as a sync.  This may be done with repeat
+       * packets and normal samples.  But if a reset has been sent
+       * then we assume that a long time has passed, so we send a
+       * space with the maximum time value.
+       */
+
+      sample = LIRC_SPACE(LIRC_VALUE_MASK);
+      rcinfo("delivering reset sync space to lirc_dev\n");
+    }
+  else if (ev.carrier_report)
+    {
+      /* Carrier reports */
+
+      sample = LIRC_FREQUENCY(ev.carrier);
+      rcinfo("carrier report (freq: %d)\n", sample);
+    }
+  else if (ev.timeout)
+    {
+      /* Packet end */
+
+      if (upper->gap)
+        {
+          return;
+        }
+
+      upper->gap = true;
+      upper->gap_start = lirc_get_timestamp() / 1000;
+      upper->gap_duration = ev.duration;
+
+      sample = LIRC_TIMEOUT(ev.duration);
+      rcinfo("timeout report (duration: %d)\n", sample);
+    }
+  else
+    {
+      /* Normal sample */
+
+      if (upper->gap)
+        {
+          upper->gap_duration += (lirc_get_timestamp() / 1000) -
+                                 upper->gap_start;
+
+          /* Cap by LIRC_VALUE_MASK */
+
+          upper->gap_duration = MIN(upper->gap_duration, LIRC_VALUE_MASK);
+          gap = LIRC_SPACE(upper->gap_duration);
+
+          flags = enter_critical_section();
+          list_for_every_safe(&upper->fh, node, tmp)
+            {
+              fh = (FAR struct lirc_fh_s *)node;
+              if (lirc_buffer_write(fh->buffer, &gap, sizeof(int)))
+                {
+                  lirc_pollnotify(fh, POLLIN | POLLRDNORM);
+                  nxsem_get_value(&fh->waitsem, &semcount);
+                  if (semcount < 1)
+                    {
+                      nxsem_post(&fh->waitsem);
+                    }
+                }
+
+              leave_critical_section(flags);
+              upper->gap = false;
+            }
+        }
+
+      sample = ev.pulse ? LIRC_PULSE(ev.duration) : LIRC_SPACE(ev.duration);
+      rcinfo("delivering %uus %d to lirc\n", ev.duration, ev.pulse ? 1 : 0);
+    }
+
+  flags = enter_critical_section();
+  list_for_every_safe(&upper->fh, node, tmp)
+    {
+      fh = (FAR struct lirc_fh_s *)node;
+      if (LIRC_IS_TIMEOUT(sample) && !fh->send_timeout_reports)
+        {
+          continue;
+        }
+
+      if (lirc_buffer_write(fh->buffer, &sample, sizeof(unsigned)))
```

Comment:
please use sizeof(unsigned int) as you did on read_mode2()

Reply:
done